packages/build-standalone/src/plugins/esbuild-measure-file-size.ts

```diff
@@ -0,0 +1,95 @@
+import recursive from 'recursive-readdir';
+import path from 'path';
+import * as fs from 'fs-extra';
+import { sync as gzipSize } from 'gzip-size';
+
+import type { ModularBuildConfig, ModularBuildContext } from '../types';
+import { canReadAsset, removeFileNameHash } from '../utils';
+import type { WorkflowPlugin } from '../workflow';
+
+export const FILE_SIZES = Symbol('FILE_SIZES');
+
+export function esbuildMeasureFileSizes<T extends ModularBuildContext>(
+  config: ModularBuildConfig,
+): WorkflowPlugin<T>[] {
+  return [
+    {
+      phase: 'validate',
+      async handler(context: T) {
+        return {
+          ...context,
+          [FILE_SIZES]: await measureBeforeBuild(
```

Comment:
Every plugin enriches the plugin chain state by returning a new copy of the context. How do we handle cases in which one plugin is dependent on another? For example, we might have a chain like:

```ts
[calculateDependenciesPlugin, buildBundlePlugin, finalizeManifestPlugin]
```

where `buildPlugin` is dependent on the context returned by `calculateDependenciesPlugin` and `finalizeManifestPlugin` is dependent on both contextes returned by `calculateDependenciesPlugin` and `buildBundlePlugin`. This makes sense in this strict order, since we hypothetically might  want to write the bundled and css dependencies (from `calculateDependenciesPlugin`) in the manifest, but also use those dependencies to bundle and write the entrypoint (generated by `buildBundlePlugin`) in the manifest.

A plugin probably doesn't want to just try accessing the context for properties it needs, since the chain would throw with an unhelpful message and that would be extremely confusing for whoever is modifying the plugin chain (an infamous example of this is Webpack, where an incorrect configuration of their plugin chain often throws with a generic "cannot access property x of undefined", which is almost always impossible to debug). Some ideas off the top of my head:

1) Validate context in the `validate` step of each plugin and throw with an helpful message if the context doesn't have certain properties. This solution has the disadvantage of being quite verbose and suffers from property clash if the plugins in the chain accidentally overwrite properties.

2) Declare dependencies and let the chain creator factory validate they are respected. This means that whoever develops Modular has to be aware of plugin id / names: changing or creating a new one means manually looking for conflicts and synchronizing existing dependency declarations to align with the new name. IDK how important is it now, but if we want community plugins we also have to be aware of versioning.

What do you guys think?